picard/tagger.py

```diff
@@ -511,81 +545,93 @@ def add_files(self, filenames, target=None):
+                self._pending_files[file.filename] = (file, target)
+                self._pending_files_process_queue.put(file.filename)
 
-    def add_directory(self, path):
-        if config.setting['recursively_add_files']:
-            self._add_directory_recursive(path)
-        else:
-            self._add_directory_non_recursive(path)
+    def _file_load_finished_thread(self):
+        while True:
+            filename, metadata = self._pending_files_results_queue.get()
 
-    def _add_directory_recursive(self, path):
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        walk = os.walk(path)
+            if self.stopping:
+                break
 
-        def get_files():
-            try:
-                root, dirs, files = next(walk)
-                if ignore_hidden:
-                    dirs[:] = [d for d in dirs if not is_hidden(os.path.join(root, d))]
-            except StopIteration:
-                return None
+            if filename in self._pending_files:
+                file, target = self._pending_files.pop(filename)
+
+                error = not metadata or None
+                thread.to_main(file._loading_finished, result=metadata, error=error, callback=None)
+
+                time.sleep(0)
+                self._file_loaded_queue.put((file, target))
+
+        # Append empty message to kill _file_loaded_thread
+        self._file_loaded_queue.put((None, None))
+
+    def _file_loaded_thread(self):
+        while True:
+            file, target = self._file_loaded_queue.get()
+
+            if self.stopping:
+                break
+
+            thread.to_main(self._file_loaded, file, target)
+            if not self._pending_files_results_queue.empty():
+                time.sleep(0.1)
             else:
-                number_of_files = len(files)
-                if number_of_files:
-                    mparms = {
-                        'count': number_of_files,
-                        'directory': root,
-                    }
-                    log.debug("Adding %(count)d files from '%(directory)r'" %
-                              mparms)
-                    self.window.set_statusbar_message(
-                        ngettext(
-                            "Adding %(count)d file from '%(directory)s' ...",
-                            "Adding %(count)d files from '%(directory)s' ...",
-                            number_of_files),
-                        mparms,
-                        translate=None,
-                        echo=None
-                    )
-                return (os.path.join(root, f) for f in files)
-
-        def process(result=None, error=None):
-            if result:
-                if error is None:
-                    self.add_files(result)
-                thread.run_task(get_files, process)
-
-        process(True, False)
-
-    def _add_directory_non_recursive(self, path):
+                time.sleep(0.01)
```

Comment:
Those `time.sleep()` calls deserve a comment, at least to explain how values were chosen.

Reply:
Done.